oauth_tools/external_idp.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import abc
+import logging
+from os.path import join
+from time import sleep
+from typing import List, Optional
+
+import requests
+from lightkube import Client, KubeConfig, codecs
+from lightkube.core.exceptions import ApiError, ObjectDeleted
+from lightkube.resources.apps_v1 import Deployment
+from lightkube.resources.core_v1 import Namespace, Pod, Service
+from pytest_operator.plugin import OpsTest
+from requests.exceptions import RequestException
+
+from oauth_tools.constants import (
+    DEX_CLIENT_ID,
+    DEX_CLIENT_SECRET,
+    DEX_MANIFESTS,
+    EXTERNAL_USER_EMAIL,
+    EXTERNAL_USER_PASSWORD,
+    KUBECONFIG,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalIdpManager(abc.ABC):
+    """Abstract class for managing lifecycle for an external IdP."""
+
+    @property
+    def client_id(self) -> str:
+        """The client_id of a registered client."""
+        ...
+
+    @property
+    def client_secret(self) -> str:
+        """The client_sercet of a registered client."""
+        ...
+
+    @property
+    def user_email(self) -> str:
+        """The test user's email."""
+        ...
+
+    @property
+    def user_password(self) -> str:
+        """The test user's password."""
+        ...
+
+    @property
+    def issuer_url(self) -> str:
+        """The provider's issuer URL."""
+        ...
+
+    @abc.abstractmethod
+    def create_idp_service(self) -> None:
+        """Deploy and configure the idp service."""
+        ...
+
+    @abc.abstractmethod
+    def wait_until_is_ready(self) -> None:
+        """Wait until the idp service is ready."""
+        ...
+
+    @abc.abstractmethod
+    def remove_idp_service(self) -> None:
+        """Remove and clean up the idp service."""
+        ...
+
+    @abc.abstractmethod
+    def update_redirect_uri(self, redirect_uri: str) -> None:
+        """Update the registered client's redirect_uri."""
+        ...
+
+
+class DexIdpManager(ExternalIdpManager):
+    """Class for managing lifecycle for an external Dex IdP."""
+
+    client_id = DEX_CLIENT_ID
+    client_secret = DEX_CLIENT_SECRET
+    user_email = EXTERNAL_USER_EMAIL
+    user_password = EXTERNAL_USER_PASSWORD
+    _namespace = "dex"
+
+    def __init__(self, ops_test: OpsTest, client: Optional[Client] = None):
```

Comment:
Just a **NONBLOCKING** question: from what I understand, `DexIdpManager` does not have to be passed with this argument `ops_test`, correct? If we would like to keep the resources being destroyed, we probably could just send in a flag from the caller.